In [27]:
import pandas as pd
# pd.set_option('max columns', 20)

# 1. IT Salary Survey EU 2018
Read the CSV file using pandas

In [28]:
df_2018 = pd.read_csv("IT Salary Survey EU 2018.csv")

## 1. Understanding the data
* **Shape** 
    * ```df.shape``` - tuple that tells us the size of our data frame. Row * Column (Like the order of a matrix)
* **Row Analysis** 
    * ```df.head()``` - function that shows the first 5 rows
    * ```df.tail()``` - function that shows the last 5 rows
* **Column Analysis** 
    * ```df.columns``` - used to list all the columns
    * ```df.dtypes``` - determines the data types of all the columns
    * ```df.describe``` - my favorite! Gives small numeric data or statistics about the columns

### 1. Shape Analysis
Find out the shape of the data frame using ```df.shape```. Note that this is not a function but a pre-stored tuple.

In [29]:
print(df_2018.shape)

(765, 14)


### 2. Row analysis

Print the first and last 5 lines of the csv data to know the size of the data we're working with

+ First five:

In [30]:
# First 5
print(df_2018.head())

             Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary  Salary one year ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   
3                  6.0     Senior         78000.0              55000.0   
4                 10.0     Senior         69000.0              60000.0   

   Salary two years ago Are you getting any Stock Options?  \
0               68

+ Last 5:

In [31]:
# Last 5
print(df_2018.tail())

               Timestamp   Age Gender     City               Position  \
760  03/06/2020 20:12:51  40.0      M     Köln  Java Developer junior   
761  28/07/2020 04:03:13   NaN      M     Köln      E.g. C# Developer   
762  28/07/2020 04:03:20   NaN      M     Köln      E.g. C# Developer   
763  26/08/2020 09:06:44   NaN    NaN      NaN                    NaN   
764  21/09/2020 01:47:10  31.0      F  München                    Pm    

     Years of experience Your level  Current Salary  Salary one year ago  \
760                  1.0     Junior         44000.0              40000.0   
761                  1.0     Junior         45000.0              40000.0   
762                  1.0     Junior         45000.0              40000.0   
763                  NaN        NaN             NaN                  NaN   
764                 10.0     Senior        110000.0                  NaN   

     Salary two years ago Are you getting any Stock Options?  \
760               40000.0               

Since our first entry is at index 0, we have 764 entries in total. This concurs with the df.shape we ran

### 2. Column analysis
+ As we can see from the shape, we have 14 columns. What are they?

In [32]:
print(df_2018.columns)

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Salary one year ago',
       'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type'],
      dtype='object')


+ Determine the data type of the columns

In [33]:
print(df_2018.dtypes)

Timestamp                              object
Age                                   float64
Gender                                 object
City                                   object
Position                               object
Years of experience                   float64
Your level                             object
Current Salary                        float64
Salary one year ago                   float64
Salary two years ago                  float64
Are you getting any Stock Options?     object
Main language at work                  object
Company size                           object
Company type                           object
dtype: object


+ Get insights on the numeric data/statistics about the columns

In [34]:
df_2018.describe()

Age  Years of experience  Current Salary  Salary one year ago  \
count  672.000000           732.000000      750.000000           596.000000   
mean    32.183036             8.548497    68381.765333         62187.278523   
std      5.107268             4.729557    21196.306557         20163.008663   
min     21.000000             0.000000    10300.000000         10001.000000   
25%     29.000000             5.000000    57000.000000         52000.000000   
50%     32.000000             8.000000    65000.000000         60000.000000   
75%     35.000000            11.000000    75000.000000         70000.000000   
max     60.000000            38.000000   200000.000000        200000.000000   

       Salary two years ago  
count            463.000000  
mean           58013.475162  
std            20413.048908  
min            10001.000000  
25%            48000.000000  
50%            56000.000000  
75%            67000.000000  
max           150000.000000

## 2. Preparing the Data (Data Cleaning)
This steps involves:
+ Dropping irrelevant columns and rows
+ Identifying duplicated columns
+ Renaming columns
+ Feature Creation

Running ```df.head()``` just to remember our columns and how our data set looks like

In [35]:
df_2018.head()

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary  Salary one year ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   
3                  6.0     Senior         78000.0              55000.0   
4                 10.0     Senior         69000.0              60000.0   

   Salary two years ago Are you getting any Stock Options?  \
0               68000.0                                 No   
1               55000.0                                 No   
2               54000.0                                 No   
3               45000.0                                Yes   
4               52000.0                                 No   

  Main language at work Company size   Company type  
0               Deutsch     100-1000        Product  
1               Deutsch       50-100        Product  
2               Deutsch        1000+        Product  
3               English        1000+        Product  
4               English     100-1000  Ecom retailer

### i. Dropping irrelevant rows or columns
There are two ways of doing this
+ Using the drop function
+ List subsetting (preferred)

#### The drop function
Say that we find the "Are you getting any stock options?" data irrelevant. We can remove it using the drop function ```df.columns.drop()```

In [36]:
df_2018.columns.drop("Are you getting any Stock Options?")

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Salary one year ago',
       'Salary two years ago', 'Main language at work', 'Company size',
       'Company type'],
      dtype='object')

If you want to drop more than one column, you can do so by using a list.

In [37]:
df_2018.columns.drop(["Salary one year ago", "Salary two years ago"])

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type'],
      dtype='object')

Note how "Are you getting any stock options?" is back yet we had dropped it. If you wanted to analyze data without that column, you should do it in the same cell you dropped it.

#### List Subsetting
Steps:
+ Run ```df.columns``` to see all the columns
+ Copy and paste the list of columns into a cell
+ Comment out the columns you don't want and run ```df[[list of columns]]``` (The two square brackets are **extremely** important)

In [38]:
df_2018.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Salary one year ago',
       'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type'],
      dtype='object')

In [39]:
df_2018[['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', #'Salary one year ago',
       #'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type']]

Timestamp   Age Gender     City                       Position  \
0    14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1    14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2    14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3    14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4    14/12/2018 12:50:31  39.0      M  München                    UX Designer   
..                   ...   ...    ...      ...                            ...   
760  03/06/2020 20:12:51  40.0      M     Köln          Java Developer junior   
761  28/07/2020 04:03:13   NaN      M     Köln              E.g. C# Developer   
762  28/07/2020 04:03:20   NaN      M     Köln              E.g. C# Developer   
763  26/08/2020 09:06:44   NaN    NaN      NaN                            NaN   
764  21/09/2020 01:47:10  31.0      F  München                            Pm    

     Years of experience Your level  Current Salary Main language at work  \
0                   11.0     Senior         77000.0               Deutsch   
1                    8.0     Senior         65000.0               Deutsch   
2                   10.0     Senior         88000.0               Deutsch   
3                    6.0     Senior         78000.0               English   
4                   10.0     Senior         69000.0               English   
..                   ...        ...             ...                   ...   
760                  1.0     Junior         44000.0               Deutsch   
761                  1.0     Junior         45000.0               Deutsch   
762                  1.0     Junior         45000.0               Deutsch   
763                  NaN        NaN             NaN                   NaN   
764                 10.0     Senior        110000.0               English   

    Company size   Company type  
0       100-1000        Product  
1         50-100        Product  
2          1000+        Product  
3          1000+        Product  
4       100-1000  Ecom retailer  
..           ...            ...  
760        10-50        Product  
761        10-50        Product  
762        10-50        Product  
763          NaN            NaN  
764        1000+        Product  

[765 rows x 11 columns]

However, just like using the drop function, this only applies to the current cell and doesn't transcend to later cells. This can be proven by running ```df.head()``` in a new cell.

In [40]:
df_2018.head()

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary  Salary one year ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   
3                  6.0     Senior         78000.0              55000.0   
4                 10.0     Senior         69000.0              60000.0   

   Salary two years ago Are you getting any Stock Options?  \
0               68000.0                                 No   
1               55000.0                                 No   
2               54000.0                                 No   
3               45000.0                                Yes   
4               52000.0                                 No   

  Main language at work Company size   Company type  
0               Deutsch     100-1000        Product  
1               Deutsch       50-100        Product  
2               Deutsch        1000+        Product  
3               English        1000+        Product  
4               English     100-1000  Ecom retailer

However, to keep working with the new subsetted data in new cells, you'll need to reassign it to the ```df``` variable or assign it to a new variable eg. ```df_subset_1```

In [41]:
df_subset_1 = df_2018[
       ['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', #'Salary one year ago',
       #'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type']
       ]

Now we can work with this subsetted data in a new cell. To prove this, we can run ```df.shape``` or ```df.head()``` to observe the changes in the columns that were applied in the previous cell.

In [42]:
df_subset_1.shape

(765, 11)

We had 14 columns but now we have 11.

In [43]:
df_subset_1.head()

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary Main language at work  \
0                 11.0     Senior         77000.0               Deutsch   
1                  8.0     Senior         65000.0               Deutsch   
2                 10.0     Senior         88000.0               Deutsch   
3                  6.0     Senior         78000.0               English   
4                 10.0     Senior         69000.0               English   

  Company size   Company type  
0     100-1000        Product  
1       50-100        Product  
2        1000+        Product  
3        1000+        Product  
4     100-1000  Ecom retailer

In [44]:
df_subset_1.dtypes

Timestamp                 object
Age                      float64
Gender                    object
City                      object
Position                  object
Years of experience      float64
Your level                object
Current Salary           float64
Main language at work     object
Company size              object
Company type              object
dtype: object

Data Preparation also involves **changing the dtype** of necessary columns. As we can see from above, timestamp is an object yet it should be a datetime. This can be changed

In [45]:
pd.to_datetime(df_2018["Timestamp"])

0     2018-12-14 12:41:33
1     2018-12-14 12:42:09
2     2018-12-14 12:47:36
3     2018-12-14 12:50:15
4     2018-12-14 12:50:31
              ...        
760   2020-03-06 20:12:51
761   2020-07-28 04:03:13
762   2020-07-28 04:03:20
763   2020-08-26 09:06:44
764   2020-09-21 01:47:10
Name: Timestamp, Length: 765, dtype: datetime64[ns]

The code above changes the timestamp datatype of the initial dataframe into a datetime. Does this affect new dataframes generated from the initial dataframe? Only one way to find out.

In [46]:
df_subset_1.dtypes

Timestamp                 object
Age                      float64
Gender                    object
City                      object
Position                  object
Years of experience      float64
Your level                object
Current Salary           float64
Main language at work     object
Company size              object
Company type              object
dtype: object

Apparently, no it doesn't. We therefore have to change it ourselves.

In [47]:
pd.to_datetime(df_subset_1["Timestamp"])

0     2018-12-14 12:41:33
1     2018-12-14 12:42:09
2     2018-12-14 12:47:36
3     2018-12-14 12:50:15
4     2018-12-14 12:50:31
              ...        
760   2020-03-06 20:12:51
761   2020-07-28 04:03:13
762   2020-07-28 04:03:20
763   2020-08-26 09:06:44
764   2020-09-21 01:47:10
Name: Timestamp, Length: 765, dtype: datetime64[ns]

During data preparation we may need to **rename the columns**. This might be the case when our column names have special characters, spaces or are in the wrong case. In our data, we'll want to remove the spaces in the column names and replace them with underscores.

In [48]:
for column in df_2018.columns:
    if " " in column:
        df_2018 = df_2018.rename(columns = {column: column.replace(" ", "_")})

df_2018.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years_of_experience',
       'Your_level', 'Current_Salary', 'Salary_one_year_ago',
       'Salary_two_years_ago', 'Are_you_getting_any_Stock_Options?',
       'Main_language_at_work', 'Company_size', 'Company_type'],
      dtype='object')

This could also be done by renaming each column, one by one using a dictionary, but I wanted to be fancy.

The non-fancy way:
```python
df_2018 = df_2018.rename(columns = 
    {"Years of experience": "Years_of_experience"
    "Your level": "Your_level"
    # and so on
    })
```

Note: You should have renamed this before you subsetted the data because, the columns of your subset still need to be renamed. See:

In [49]:
df_subset_1.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Main language at work', 'Company size',
       'Company type'],
      dtype='object')

Now you'll have to subset the data all over again

In [50]:
df_subset_1 = df_2018[
    ['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years_of_experience',
       'Your_level', 'Current_Salary', 'Salary_one_year_ago',
       'Salary_two_years_ago', #'Are_you_getting_any_Stock_Options?',
       'Main_language_at_work', 'Company_size', 'Company_type']
]

df_subset_1.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years_of_experience',
       'Your_level', 'Current_Salary', 'Salary_one_year_ago',
       'Salary_two_years_ago', 'Main_language_at_work', 'Company_size',
       'Company_type'],
      dtype='object')

To confirm that the data is correctly subsetted, we can run ```df.head()```

In [51]:
df_subset_1.head()

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years_of_experience Your_level  Current_Salary  Salary_one_year_ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   
3                  6.0     Senior         78000.0              55000.0   
4                 10.0     Senior         69000.0              60000.0   

   Salary_two_years_ago Main_language_at_work Company_size   Company_type  
0               68000.0               Deutsch     100-1000        Product  
1               55000.0               Deutsch       50-100        Product  
2               54000.0               Deutsch        1000+        Product  
3               45000.0               English        1000+        Product  
4               52000.0               English     100-1000  Ecom retailer

Now that we are comfortable with the names of our columns and the subsetted data, we can now **look for missing values**. Running the ```df.isna()``` command shows us where we have null values using booleans. True means we have a null value there and false means it isn't null.

In [52]:
df_subset_1.isna()

Timestamp    Age  Gender   City  Position  Years_of_experience  \
0        False  False   False  False     False                False   
1        False  False   False  False     False                False   
2        False  False   False  False     False                False   
3        False  False   False  False     False                False   
4        False  False   False  False     False                False   
..         ...    ...     ...    ...       ...                  ...   
760      False  False   False  False     False                False   
761      False   True   False  False     False                False   
762      False   True   False  False     False                False   
763      False   True    True   True      True                 True   
764      False  False   False  False     False                False   

     Your_level  Current_Salary  Salary_one_year_ago  Salary_two_years_ago  \
0         False           False                False                 False   
1         False           False                False                 False   
2         False           False                False                 False   
3         False           False                False                 False   
4         False           False                False                 False   
..          ...             ...                  ...                   ...   
760       False           False                False                 False   
761       False           False                False                 False   
762       False           False                False                 False   
763        True            True                 True                  True   
764       False           False                 True                  True   

     Main_language_at_work  Company_size  Company_type  
0                    False         False         False  
1                    False         False         False  
2                    False         False         False  
3                    False         False         False  
4                    False         False         False  
..                     ...           ...           ...  
760                  False         False         False  
761                  False         False         False  
762                  False         False         False  
763                   True          True          True  
764                  False         False         False  

[765 rows x 13 columns]

As you can see, line 763 and 764 have null values as they contain True. Line 763 in particular is wanting as it only has one false value. It's like the person only answered one question and got tired. However, this is a bit overwhelming to look at. Furthermore, we can't see all our 765 rows, so we really can't tell how many null values we have in there.

To get a comprehensive summary of all the null values in each column, we use append the ```.sum()``` function to the ```.isna()``` function.

In [53]:
df_subset_1.isna().sum()

Timestamp                  0
Age                       93
Gender                    14
City                      29
Position                  28
Years_of_experience       33
Your_level                22
Current_Salary            15
Salary_one_year_ago      169
Salary_two_years_ago     302
Main_language_at_work     15
Company_size              15
Company_type              35
dtype: int64

This shows us that we have 0 null values in the Timestamp column, 93 null values in the Age column, 14 in Gender, 29 in city and so on. Going back to line 763, this person didn't answer any question to begin with, the only present value is the timestamp. 

We can also see the columns with the highest null values are "Salary_two_years_ago" and "Salary_one_year_ago" respectively. This might be because these people must have forgotten what they were earning a year or two earlier. The next column with the highest null value is "Age" and this may be because people don't like disclosing their age.

Similarly, we might need to see if we have duplicated rows. We do this by running ```df.duplicated()```.

In [57]:
df_subset_1.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
760    False
761    False
762    False
763    False
764    False
Length: 765, dtype: bool

Just like ```.isna()```, ```duplicated()``` produces boolean values where True in this case shows us a duplicated row.

We can use the ```df.loc[]``` to locate each row that is duplicated. "loc" is just short for location indexer and it helps us get the location of all the True values. To learn more on "loc", run ```help(df.loc)```

In [58]:
help(df_subset_1.loc)

Help on _LocIndexer in module pandas.core.indexing object:

class _LocIndexer(_LocationIndexer)
 |  Access a group of rows and columns by label(s) or a boolean array.
 |  
 |  ``.loc[]`` is primarily label based, but may also be used with a
 |  boolean array.
 |  
 |  Allowed inputs are:
 |  
 |  - A single label, e.g. ``5`` or ``'a'``, (note that ``5`` is
 |    interpreted as a *label* of the index, and **never** as an
 |    integer position along the index).
 |  - A list or array of labels, e.g. ``['a', 'b', 'c']``.
 |  - A slice object with labels, e.g. ``'a':'f'``.
 |  
 |    .. warning:: Note that contrary to usual python slices, **both** the
 |        start and the stop are included
 |  
 |  - A boolean array of the same length as the axis being sliced,
 |    e.g. ``[True, False, True]``.
 |  - An alignable boolean Series. The index of the key will be aligned before
 |    masking.
 |  - An alignable Index. The Index of the returned selection will be the input.
 |  - A ``callable`

In [59]:
df_subset_1.loc[df_subset_1.duplicated()]

Empty DataFrame
Columns: [Timestamp, Age, Gender, City, Position, Years_of_experience, Your_level, Current_Salary, Salary_one_year_ago, Salary_two_years_ago, Main_language_at_work, Company_size, Company_type]
Index: []

In our data, we don't have duplicated rows.

We can also do this for a subset of columns to see if we have duplicates

In [63]:
df_subset_1.duplicated(subset=["Position"])

0      False
1      False
2      False
3      False
4      False
       ...  
760    False
761    False
762     True
763     True
764    False
Length: 765, dtype: bool

There are duplicates. Let's locate them.

In [64]:
df_subset_1.loc[df_subset_1.duplicated(subset=["Position"])]

Timestamp   Age Gender     City            Position  \
10   14/12/2018 12:56:28  40.0      M   Berlin   Software Engineer   
17   14/12/2018 13:10:26  24.0      M  Hamburg   Software Engineer   
18   14/12/2018 13:16:09  44.0      M   Berlin  Frontend Developer   
26   14/12/2018 13:32:27  39.0      M  München  Frontend Developer   
31   14/12/2018 13:40:40  35.0      F  München   Software Engineer   
..                   ...   ...    ...      ...                 ...   
757  11/03/2020 23:23:19   NaN    NaN      NaN                 NaN   
758  31/05/2020 16:28:15   NaN      M   Berlin                 NaN   
759  31/05/2020 16:28:23   NaN      M   Berlin                 NaN   
762  28/07/2020 04:03:20   NaN      M     Köln   E.g. C# Developer   
763  26/08/2020 09:06:44   NaN    NaN      NaN                 NaN   

     Years_of_experience Your_level  Current_Salary  Salary_one_year_ago  \
10                  18.0     Senior         70000.0              57500.0   
17                   4.0     Middle         60000.0              47000.0   
18                  17.0     Senior         62500.0              62500.0   
26                  15.0     Senior         77000.0              74000.0   
31                  12.0     Senior         84000.0              84000.0   
..                   ...        ...             ...                  ...   
757                  NaN        NaN             NaN                  NaN   
758                  NaN        NaN             NaN                  NaN   
759                  NaN        NaN             NaN                  NaN   
762                  1.0     Junior         45000.0              40000.0   
763                  NaN        NaN             NaN                  NaN   

     Salary_two_years_ago Main_language_at_work Company_size Company_type  
10                54000.0               English     100-1000      Product  
17                    NaN               English        1000+      Product  
18                54000.0               Deutsch        10-50       Agency  
26                    NaN               English     100-1000      Product  
31                84000.0               English     100-1000      Startup  
..                    ...                   ...          ...          ...  
757                   NaN                   NaN          NaN          NaN  
758                   NaN                   NaN          NaN          NaN  
759                   NaN                   NaN          NaN          NaN  
762               40000.0               Deutsch        10-50      Product  
763                   NaN                   NaN          NaN          NaN  

[367 rows x 13 columns]

I'm also seeing a lot of NaN values. I know this means Not a Number in JavaScript, but what does it mean here? I see they mostly appear towards the bottom of the page, so let's run tail to get a better understanding.

In [65]:
df_subset_1.tail(10)

Timestamp   Age Gender     City               Position  \
755  06/01/2020 13:38:05  30.0      M  München     Softwareentwickler   
756  05/03/2020 16:50:43  52.0      M     Köln            IT Engineer   
757  11/03/2020 23:23:19   NaN    NaN      NaN                    NaN   
758  31/05/2020 16:28:15   NaN      M   Berlin                    NaN   
759  31/05/2020 16:28:23   NaN      M   Berlin                    NaN   
760  03/06/2020 20:12:51  40.0      M     Köln  Java Developer junior   
761  28/07/2020 04:03:13   NaN      M     Köln      E.g. C# Developer   
762  28/07/2020 04:03:20   NaN      M     Köln      E.g. C# Developer   
763  26/08/2020 09:06:44   NaN    NaN      NaN                    NaN   
764  21/09/2020 01:47:10  31.0      F  München                    Pm    

     Years_of_experience Your_level  Current_Salary  Salary_one_year_ago  \
755                 10.0     Middle         75000.0              50000.0   
756                 30.0     Middle         63000.0              80000.0   
757                  NaN        NaN             NaN                  NaN   
758                  NaN        NaN             NaN                  NaN   
759                  NaN        NaN             NaN                  NaN   
760                  1.0     Junior         44000.0              40000.0   
761                  1.0     Junior         45000.0              40000.0   
762                  1.0     Junior         45000.0              40000.0   
763                  NaN        NaN             NaN                  NaN   
764                 10.0     Senior        110000.0                  NaN   

     Salary_two_years_ago Main_language_at_work Company_size Company_type  
755                   NaN               Deutsch     100-1000       Agency  
756               80000.0               English        1000+       Agency  
757                   NaN                   NaN          NaN          NaN  
758                   NaN                   NaN          NaN          NaN  
759                   NaN                   NaN          NaN          NaN  
760               40000.0               Deutsch        10-50      Product  
761               40000.0               Deutsch        10-50      Product  
762               40000.0               Deutsch        10-50      Product  
763                   NaN                   NaN          NaN          NaN  
764                   NaN               English        1000+      Product

When I investigate these NaN values in the CSV, they are actually empty values/null values. We might want to remove the rows where we have nothing completely.